src/main/drivers/io_preinit.c

```diff
@@ -0,0 +1,122 @@
+/*
+ * This file is part of Cleanflight and Betaflight.
+ *
+ * Cleanflight and Betaflight are free software. You can redistribute
+ * this software and/or modify this software under the terms of the
+ * GNU General Public License as published by the Free Software
+ * Foundation, either version 3 of the License, or (at your option)
+ * any later version.
+ *
+ * Cleanflight and Betaflight are distributed in the hope that they
+ * will be useful, but WITHOUT ANY WARRANTY; without even the implied
+ * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this software.
+ *
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <stdbool.h>
+#include <stdint.h>
+
+#include "platform.h"
+
+#include "drivers/io.h"
+#include "drivers/resource.h"
+#include "drivers/system.h"
+
+#include "drivers/flash.h"
+#include "drivers/max7456.h"
+#include "drivers/rx/rx_spi.h"
+#include "drivers/sdcard.h"
+
+#include "pg/flash.h"
+#include "pg/max7456.h"
+#include "pg/rx_spi.h"
+#include "pg/sdcard.h"
+
+#include "sensors/initialisation.h"
+
+typedef struct ioPreInit_s {
+    ioTag_t iotag;
+    uint8_t iocfg;
+    uint8_t init;
+} ioPreInit_t;
+
+static ioPreInit_t ioPreInitArray[IO_PREINIT_COUNT];
+static int ioPreInitCount = 0;
+
+void ioPreInitRegister(ioTag_t iotag, uint8_t iocfg, uint8_t init)
+{
+    if (!iotag) {
+        return;
+    }
+
+    if (ioPreInitCount == IO_PREINIT_COUNT) {
+        indicateFailure(FAILURE_DEVELOPER, 5);
+        return;
+    }
+
+    ioPreInitArray[ioPreInitCount].iotag = iotag;
+    ioPreInitArray[ioPreInitCount].iocfg = iocfg;
+    ioPreInitArray[ioPreInitCount].init = init;
+    ++ioPreInitCount;
+}
+
+static void ioPreInit(ioPreInit_t *preinit, int index)
+{
+    IO_t io = IOGetByTag(preinit->iotag);
+    IOInit(io, OWNER_PREINIT, RESOURCE_INDEX(index));
+    IOConfigGPIO(io, preinit->iocfg);
+    if (preinit->init) {
+        IOHi(io);
+    } else {
+        IOLo(io);
+    }
+}
+
+void ioPreInitAll(void)
+{
+    sensorsPreInit();
```

Comment:
Not 100% sure if we shouldn't put all `_preInit` into here instead of subgrouping into `sensorsPreInit()`.

Reply:
It was motivated by the presence of `sensorsAutodetect`, which groups sensor detections without no hard relationship between them. Thoughts?

Reply:
Yeah, I guess since we're already using this pattern we can as well stick to it with `sensorsPreInit()`.